### PR TITLE
Propose a solution for #3172, initially covering long URL on mobile phone screens

### LINKED ISSUE #3172: Wrap long lines when displaying message (text size gets too small)
K9 doesn't warp links, which makes some e-mails very wide:
screenshot_20180208-131048
***
compared to:
screenshot_20180208-131057

### PR DESCRIPTION
**Pull request by joohnq**
Fixes #3172

This update improves readability by ensuring that long words and links in email messages wrap properly in the WebView, avoiding horizontal scrolling. In DisplayHtml.kt, the cssStylePre method now includes CSS rules for <a> tags (word-break and overflow-wrap). Additionally, the HTML and CSS generation methods have been refactored to use multiline string literals for better message clarity.

### CODE REVIEW

File: legacy/ui/legacy/src/main/java/com/fsck/k9/ui/messageview/MessageContainerView.kt [hunk 1]
```diff
@@ -433,8 +433,8 @@ class MessageContainerView(context: Context, attrs: AttributeSet?) :
             }
         }
 
-        val textToDisplay = messageText
-            ?: displayHtml.wrapStatusMessage(context.getString(R.string.webview_empty_message))
+        val textToDisplay =
+            displayHtml.wrapStatusMessage(messageText ?: context.getString(R.string.webview_empty_message))
```
Comment (rafaeltonholo) [inline legacy/ui/legacy/src/main/java/com/fsck/k9/ui/messageview/MessageContainerView.kt]: The `displayHtml.wrapStatusMessage(CharSequence)` wraps the whole content within a div with an inline style of `text-align:center; color: grey`, thus causing all the text within the message to be centre-aligned and gray.

File: legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt [hunk 2]
```diff
@@ -16,30 +16,53 @@ class DisplayHtml(private val settings: HtmlSettings) : HtmlHeadProvider {
...
+                <body>
+                    $messageContent
+                </body>
+            </html>
+        """.trimIndent()
     }
 
     /**
-     * Dynamically generate a CSS style for `<pre>` elements.
+     * Dynamically generate a CSS style for <pre> elements.
```
Comment (rafaeltonholo) [inline legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt]: Could you please revert this change? The backtick is important to help a better rendering of the documentation by either inspecting it via auto complete or by turning on the Reader mode on Android Studio.

File: legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt [hunk 3]
```diff
@@ -16,30 +16,53 @@ class DisplayHtml(private val settings: HtmlSettings) : HtmlHeadProvider {
...
      *
      * The style incorporates the user's current preference setting for the font family used for plain text messages.
      *
-     * @return A `<style>` element that can be dynamically included in the HTML `<head>` element when messages are
+     * @return A <style> element that can be dynamically included in the HTML <head> element when messages are
```
Comment (rafaeltonholo) [inline legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt]: Please revert this change.

File: legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt [hunk 4]
```diff
@@ -16,30 +16,53 @@ class DisplayHtml(private val settings: HtmlSettings) : HtmlHeadProvider {
...
      * displayed.
      */
     private fun cssStylePre(): String {
         val font = if (settings.useFixedWidthFont) "monospace" else "sans-serif"
 
-        return "<style type=\"text/css\"> pre." + EmailTextToHtml.K9MAIL_CSS_CLASS +
-            " {white-space: pre-wrap; word-wrap:break-word; " +
-            "font-family: " + font + "; margin-top: 0px}</style>"
+        return """
```
Comment (rafaeltonholo) [inline legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt:77]: As we are improving the readability of this html code, could you please also annotate the method with `@Language("HTML")`? It will enable syntax highlight on Android Studio/IntelliJ

File: legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt [hunk 5]
```diff
@@ -16,30 +16,53 @@ class DisplayHtml(private val settings: HtmlSettings) : HtmlHeadProvider {
...
+                    word-break: break-word;
+                    overflow-wrap: break-word;
+                }
+            </style>
+        """.trimIndent()
     }
 
     private fun cssStyleSignature(): String {
-        return """<style type="text/css">.k9mail-signature { opacity: 0.5 }</style>"""
+        return """
```
Comment (rafaeltonholo) [inline legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt:91]: Could you please also annotate the method with `@Language("HTML")`?

File: legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt [hunk 6]
```diff
@@ -16,30 +16,53 @@ class DisplayHtml(private val settings: HtmlSettings) : HtmlHeadProvider {
 
     fun wrapMessageContent(messageContent: CharSequence): String {
         // Include a meta tag so the WebView will not use a fixed viewport width of 980 px
-        return "<html dir=\"auto\"><head><meta name=\"viewport\" content=\"width=device-width\"/>" +
-            cssStylePre() +
-            "</head><body>" +
-            messageContent +
-            "</body></html>"
+        return """
```
Comment (rafaeltonholo) [inline legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt:32]: Could you please also annotate the method with `@Language("HTML")`?

File: legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt [hunk 7]
```diff
@@ -16,30 +16,53 @@ class DisplayHtml(private val settings: HtmlSettings) : HtmlHeadProvider {
...
-            "font-family: " + font + "; margin-top: 0px}</style>"
+        return """
+            <style type="text/css">
+                pre.${EmailTextToHtml.K9MAIL_CSS_CLASS} {
+                    white-space: pre-wrap;
+                    word-wrap: break-word;
+                    font-family: $font;
+                    margin-top: 0px;
+                }
+                a {
```
Comment (rafaeltonholo) [inline legacy/core/src/main/java/com/fsck/k9/message/html/DisplayHtml.kt]: I believe we could be more aggressive and apply the `word-break` and `overflow-wrap` properties to all text, rather than just the `a` tag. This approach is similar to what other email providers do.

**Review by rafaeltonholo — CHANGES_REQUESTED** [review]
Hi @joohnq,

Thank you for your contribution! There are just a few comments that need to be addressed before we can merge these changes.

Additionally, to fully resolve the issue, please make sure to add `initial-scale=1.0` to the `<meta name="viewport" />`. This adjustment will ensure that the email's content maintains its original scale without resizing, which helps prevent the text-shrink problem.

Here’s the updated code:

```diff
override val headHtml: String
    get() {
--        return """<meta name="viewport" content="width=device-width"/>""" +
++        return """<meta name="viewport" content="width=device-width, initial-scale=1.0"/>""" +
            cssStylePre() +
            cssStyleSignature()
    }
```

Thank you!

**Comment by joohnq** [comment]
All right, @rafaeltonholo. Do you think it’s better to send a single commit to fix everything, or separate smaller commits for each observation?

**Comment by rafaeltonholo** [comment]
> All right, @rafaeltonholo. Do you think it’s better to send a single commit to fix everything, or separate smaller commits for each observation?

I would rebase the first commit of this PR and include the changes there, as they are part of the same change, but you can proceed with what is easier for you. As soon as it follows our [commits guidelines](https://thunderbird.github.io/thunderbird-android/docs/latest/contributing/git-commit-guide.html), it is all good :)

**Comment by joohnq** [comment]
> Hi @joohnq,
> 
> Thank you for your contribution! There are just a few comments that need to be addressed before we can merge these changes.
> 
> Additionally, to fully resolve the issue, please make sure to add `initial-scale=1.0` to the `<meta name="viewport" />`. This adjustment will ensure that the email's content maintains its original scale without resizing, which helps prevent the text-shrink problem.
> 
> Here’s the updated code:
> 
> ```diff
> override val headHtml: String
>     get() {
> --        return """<meta name="viewport" content="width=device-width"/>""" +
> ++        return """<meta name="viewport" content="width=device-width, initial-scale=1.0"/>""" +
>             cssStylePre() +
>             cssStyleSignature()
>     }
> ```
> 
> Thank you!

Most of the updates mentioned are already in place. As for the initial-scale=1.0, I’ve tested it with a variety of emails — including some with content that exceeds the screen width — and in practice, not setting the scale actually provided better results in some cases.

The line breaking handled by the CSS seems to cover most edge cases and improves readability. That said, I’m open to adjusting it if you think it’s strictly necessary for consistency or device compatibility.

**Comment by joohnq** [comment]
### With `initial-scale=1.0`

These screenshots show the behavior when `initial-scale=1.0` is applied:

<table>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td colspan="3" align="center"></td>
  </tr>
</table>

---

### Without `initial-scale=1.0`

These screenshots show the behavior without setting the `initial-scale` parameter:

<table>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td colspan="3" align="center"></td>
  </tr>
</table>

**Comment by joohnq** [comment]
Well, I updated the code based on what we discussed @rafaeltonholo , but I'll run more tests with additional scenarios.

**Review by rafaeltonholo — APPROVED** [review]
LGTM! Thank you for this contribution and for the dedication to checking the edge cases!
